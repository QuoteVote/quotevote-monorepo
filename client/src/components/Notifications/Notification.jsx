import React from 'react'
import { Grid } from '@mui/material'
import Typography from '@mui/material/Typography'
import { makeStyles } from '@mui/material/styles'
import { Skeleton } from '@mui/lab'
import List from '@mui/material/List'
import ListItem from '@mui/material/ListItem'
import ListItemText from '@mui/material/ListItemText'
import ListItemAvatar from '@mui/material/ListItemAvatar'
import PropTypes from 'prop-types'
import Divider from '@mui/material/Divider'
import IconButton from '@mui/material/IconButton'
import Tooltip from '@mui/material/Tooltip'
import LaunchIcon from '@mui/icons-material/Launch'
import { useHistory } from 'react-router-dom'
import NotificationLists from './NotificationLists'
import { useMobileDetection } from '../../utils/display'

const useStyles = makeStyles((theme) => ({
  root: {
    margin: 10,
    paddingRight: (props) => (props.pageView ? 30 : 0),
    backgroundColor: theme.palette.background.paper,
  },
  skeleton: {
    width: (props) => (props.pageView ? 'auto' : 350),
  },
  content: {
    width: 'inherit',
  },
  moreMenu: {
    position: 'absolute',
    top: 20,
    right: 5,
  },
}))

function Notification({
  loading, notifications, spacing = 0, pageView, setOpenPopUp,
}) {
  const isMobileDevice = useMobileDetection()
  const classes = useStyles({ pageView })
  const history = useHistory()
  const handleClick = () => {
    setOpenPopUp(false)
    history.push('/Notifications')
  }

  return (
    <Grid
      container
      direction="column"
      justify="flex-start"
      alignItems="flex-start"
      className={classes.root}
      spacing={spacing}
    >
      <Grid item>
        {!isMobileDevice && <Typography variant="h5">Notifications</Typography>}
        <Divider />
        {!pageView && (
          <Tooltip
            title="Open Notifications"
            arrow
            placement="top"
          >
            <IconButton className={classes.moreMenu} size="small" onClick={handleClick}>
              <LaunchIcon />
            </IconButton>
          </Tooltip>
        )}
      </Grid>
      <Grid item className={classes.content}>
        {loading && (
          <List className={classes.skeleton}>
            {Array.from(Array(3).keys()).map(() => (
              <ListItem>
                <ListItemAvatar>
                  <Skeleton animation="wave" variant="circle" width={40} height={40} />
                </ListItemAvatar>
                <ListItemText primary={<Skeleton animation="wave" height={10} width={pageView ? 400 : 45} style={{ marginBottom: 6 }} />} />
              </ListItem>
            ))}
          </List>
        )}
        {!loading && (
          <NotificationLists notifications={notifications} pageView={pageView} />
        )}
      </Grid>

    </Grid>
  )
}

Notification.propTypes = {
  loading: PropTypes.bool.isRequired,
  notifications: PropTypes.array.isRequired,
  spacing: PropTypes.number,
  pageView: PropTypes.bool,
  setOpenPopUp: PropTypes.func,
}

export default Notification
