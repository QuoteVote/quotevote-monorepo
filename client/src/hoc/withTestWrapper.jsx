import React from 'react'
import { PersistGate } from 'redux-persist/integration/react'
import { ApolloProvider } from '@apollo/react-hooks'
import { Provider } from 'react-redux'
import { createTheme, ThemeProvider } from '@material-ui/core/styles'
import { Router } from 'react-router-dom'
import { createBrowserHistory } from 'history'
import store, { persistor } from '../store/store'
import client from '../config/apollo'
import customTheme from '../theme'
import { AuthModalProvider } from '../Context/AuthModalContext'
const hist = createBrowserHistory()
const theme = createTheme(customTheme)

const withTestWrapper = (Component) => (props) => (
  <ApolloProvider client={client}>
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <ThemeProvider theme={theme}>
          <AuthModalProvider>
            <Router history={hist}>
              <Component {...props} />
            </Router>
          </AuthModalProvider>
        </ThemeProvider>
      </PersistGate>
    </Provider>
  </ApolloProvider>
)
export default withTestWrapper
