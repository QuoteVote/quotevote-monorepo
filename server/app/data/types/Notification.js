export const Notification = `
type Notification {
  _id: String
  userId: String
  userIdBy: String
  userBy: User
  post: Post
  notificationType: String
  label: String
  status: String
  created: Date
}
`;
