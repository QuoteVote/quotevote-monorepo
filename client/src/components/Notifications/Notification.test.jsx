import React from 'react'
import { render, act } from '@testing-library/react'
import { vi } from 'vitest'

// Component being tested
import Notification from './Notification'
import withTestWrapper from '../../hoc/withTestWrapper'

// Provide minimal props the component expects so PropTypes warnings don't appear
const TestNotification = (props) => <Notification {...props} />
const NotificationWrapper = withTestWrapper(TestNotification)

describe('Notification test -', () => {
  it('renders correctly', async () => {
    let container
    await act(async () => {
      const res = render(
        <NotificationWrapper loading={false} notifications={[]} />,
      )
      container = res.container
    })
    expect(container.firstChild).toMatchSnapshot()
  })
})
