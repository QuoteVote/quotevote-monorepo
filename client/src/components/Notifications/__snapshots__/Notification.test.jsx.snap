// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Notification test - > renders correctly 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column makeStyles-root-1 makeStyles-root-5 css-ifmrbr-MuiGrid-root"
  justify="flex-start"
>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 css-ku7d13-MuiTypography-root"
    >
      Notifications
    </h5>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-18cfrml-MuiDivider-root"
    />
    <button
      aria-label="Open Notifications"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall makeStyles-moreMenu-4 css-1jkt3ui-MuiButtonBase-root-MuiIconButton-root"
      data-mui-internal-clone-element="true"
      tabindex="0"
      type="button"
    >
      <span
        data-testid="mui-icon"
      />
    </button>
  </div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row makeStyles-content-3 css-17fpwt7-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column makeStyles-rootNoNotification-9 makeStyles-rootNoNotification-15 css-1bkjq21-MuiGrid-root"
      justify="center"
    >
      <div
        class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
      >
        <img
          alt=""
          src="/assets/ZeroNotificationsBG.png"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 css-1nmqws6-MuiTypography-root"
        >
          Relax, you don't have any alerts right now.
        </span>
      </div>
    </div>
  </div>
</div>
`;
